File: go/binary_search/binary_search_find_first.go
package binary_search

// firstInt should return first index of val in a sorted array arr, or -1 if not found
func findFirstInt(arr []int, val int) int {
	l, r := 0, len(arr)

	for l < r {
		m := (l + r) / 2

		if arr[m] < val {
			l = m + 1
		} else {
			r = m
		}
	}

	if l < len(arr) && arr[l] == val {
		return l
	}

	return -1
}

File: java/src/binary_search/BinarySearchFindFirst.java
package binary_search;

public class BinarySearchFindFirst {
	// findFirstInt should return first index of val in a sorted array arr, or -1 if not found
	public static int findFirstInt(int[] arr, int val) {
//		throw new UnsupportedOperationException("TODO");

		int l = 0, r = arr.length;
		while (l < r) {
			var m = (l + r)/2;
			if (arr[m] < val) {
				l = m+1;
			} else {
				r = m;
			}
		}

		return l < arr.length && arr[l] == val? l : -1;
	}
}
